pkg/kubelet/kubelet.go

```diff
@@ -790,11 +794,23 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 
 	tokenManager := token.NewManager(kubeDeps.KubeClient)
 
+	var clusterTrustBundleManager clustertrustbundle.Manager
```

Comment:
at minimum, this need to be feature-gated so we're not starting perpetually-failing informers against not-enabled alpha APIs in default kubelets

thinking about rollout, we may end up needing a node gate and an API gate so we can enable on the node side prior to enabling in the API server, or do some combination of API detection to see if the clustertrustbundle API is available

Reply:
I have added a new ClusterTrustBundleProjection feature gate that controls the Kubelet implementation, as well as the field-defaulting behavior on kube-apiserver.

Reply:
Actually, taking a step back...

The model I had been using to think about this originally was that the Kubelet part should be always enabled, and the control about whether or not it's in use is whether or not there are pods scheduled into the cluster that use the projected volume (which is in turn guarded by a feature gate read by kube-apiserver).  Obviously starting informers doesn't jive with that.

Maybe we could start the CTB informers when we see that we have pods scheduled on our node that need them?  Or maybe launch with a polling-based implementation that only makes calls to kube-apiserver if there is a pod that needs them?

The reason I'm thinking in this direction is that any other strategy makes it very hard to support this case without errors:

1. User enables the feature gate(s) controlling kube-apiserver and kubelet.
2. User starts using pods with clustertrustbundle projection.
3. User disables the feature gate controlling kubelet while those pods are still running.  This requires a kubelet restart, but doesn't necessarily kill the pods that are running on that node, right?

Reply:
I found [this doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#lifecycle), which seems to indicate that the way I was thinking of handling this isn't allowed (specifically, "toggling feature gates in one component should not affect other components", and "no fanout").

What does seem to work is to define two feature gates, ClusterTrustBundle and ClusterTrustBundleProjection.

To enable the feature:
1. Enable ClusterTrustBundle on all components.  Only kube-apiserver pays attention to it, and uses it to start serving ClusterTrustBundles.
2. Enable ClusterTrustBundleProjection on all components.  Kube-apiserver ignores this.  This makes kubelet switch from a no-op CTB implementation to the informer-based CTB implementation.

To disable the feature:
1. Delete all pods that use CTB projections.
2. Disable the ClusterTrustBundleProjection feature gate on all components.  If you forgot to delete all the pods, those pods will encounter errors.
3. Disable the ClusterTrustBundle feature gate on all components.

In this model, kube-apiserver never filters out or rejects pods that are trying to use ClusterTrustBundle projections.  Instead, we rely on kubelet to keep them from running.  I think if we want to have that behavior as well, we will actually need three feature gates (adding one like ClusterTrustBundleAllowProjectionsInAPI), because of the "no coordination between components is needed" language.

Also, in this model we allow the user to run into errors if they have failed to remove all running workloads that use CTB projections before they try to disable the feature.  This seems fine to me, since it's not really possible to gracefully degrade these workloads; as I noted in a different comment, your workload has to be coded to use the CTB projected volume, and if it's not present your workload will break (either silently or loudly).

Workloads that don't use CTB projections will be unaffected by any of this.